### ISSUE TITLE
Broken complexity with primitive arrays

### ISSUE THREAD

**Issue by silene** [coq/coq#12947]
Semi-persistent arrays are supposed to have an amortized complexity of O(1) for get/set, as long as they are performed on the last accessed array or an older version of it (for backtracking algorithms). Obviously, the constant factor is large, but we should be able to perform millions of array accesses per second in Coq, maybe even tens of millions. Let us try it.

```coq
Require Import BinPos Int63 PArray.

Definition foo (n : positive) :=
  let a := make 2 0 in
  let b := Pos.iter (fun b => set b 1 1) a 100000 in
  let v := get b 0 in
  (* backtrack point here, we no longer access b but only a *)
  Pos.iter (fun v => v + get a 0) v n.

Time Eval vm_compute in foo 1.       (* 0.02s *)
Time Eval vm_compute in foo 10.      (* 0.05s *)
Time Eval vm_compute in foo 100.     (* 0.37s *)
Time Eval vm_compute in foo 1000.    (* 3.56s *)
```

Millions of operations per second? No, just 300.

Why? Because `get` and `set` do not reroot the array. So, the complexity is no longer O(1); it is O(n). Moreover, it also means that the garbage collector can no longer collect discarded arrays in a backtracking algorithm, causing the memory to fill without end.

Now, the savvy user notices that `Parray` provides an explicit `reroot` function. So, perhaps we can improve performances by using it? Let us put an explicit call at the position of the backtrack point:

```coq
Definition foo (n : positive) :=
  let a := make 2 0 in
  let b := Pos.iter (fun b => set b 1 1) a 100000 in
  let v := get b 0 in
  let a := reroot a in
  Pos.iter (fun v => v + get a 0) v n.
```

Guess what? Still 300 operations per second.

Why? Because the VM does not care. More precisely, Coq decided to compile it as follows:

```coq
(* decompiled from Set Dump Bytecode *)
Definition foo (n : positive) :=
  let a := make 2 0 in
  let a1 := reroot a in
  let b := Pos.iter (fun b => set b 1 1) a 100000 in
  let v := get b 0 in
  Pos.iter (fun v => v + get a1 0) v n.
```

So, the exceedingly savvy user decides to rewrite the code as follows:

```coq
Definition foo (n : positive) :=
  let a := make 2 0 in
  let b := Pos.iter (fun b => set b 1 1) a 100000 in
  let v := get b 0 in
  Pos.iter (let a := reroot a in fun v => v + get a 0) v n.
```

Eight millions of array accesses per second. Now we are talking.

So, there is at least one bug, and potentially two:
1. `get` and `set` should be O(1), that is, they should always reroot;
2. the VM should refrain from swapping let-in constructs, to preserve the sanity of users.

**Comment by JasonGross** [comment]
Should we add this example to coq-performance-tests?

**Comment by silene** [comment]
I do not know what coq-performance-tests are, but this example is at least part of #12955.

**Comment by JasonGross** [comment]
https://github.com/coq-community/coq-performance-tests is part of the bench and also contains support for performing regression analysis of the performance of various bits of code across multiple versions of Coq (see the readme)

**Comment by silene** [comment]
I see. That would be interesting, especially since the O(1) complexity has always been under the assumption of a magical garbage collector. That said, the testcase from this issue does not look like it would show anything interesting. Indeed, it performs 100'000 `set`s, then 1 backtrack (of 100'000 `set`s), and finally n `get`s. But only the cost of the second `get` actually matters; I am doing n of them so that this cost becomes measurable. The resulting plot with respect to n would be plain affine. Indeed, hardly any memory is used during after the first `get` (even for a broken implementation). So, it would not be that interesting.

Perhaps it would be better to do n `set`s, 1 backtrack, 10'000 `get`s. It would be even better if we could interleave sets and backtracks, to simulate the behavior of an actual algorithm. But I have no idea what kind of values would be representative.

**Comment by maximedenes** [comment]
> So, there is at least one bug, and potentially two:
> 
> 1. `get` and `set` should be O(1), that is, they should always reroot;
> 2. the VM should refrain from swapping let-in constructs, to preserve the sanity of users.

After discussion with @bgregoir, the first point is nicely addressed by @silene 's #12955, and the second point should be addressed either by removing the let-in optimization, or doing an analysis to know if the let body has side effects (which includes `set` and even `get` after #12955 is merged).

**Comment by maximedenes** [comment]
> the second point should be addressed either by removing the let-in optimization, or doing an analysis to know if the let body has side effects (which includes `set` and even `get` after #12955 is merged).

@silene would you mind opening an issue about this second point, that you mentioned in this report?
